New tutorial for the Gradle BT manual (#24913)
gradle/gradle: subprojects/docs/src/docs/userguide/tutorial/part5_gradle_caching.adoc

// Copyright 2017 the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[part5_gradle_caching]]

> = Part 5: Gradle Cache

Reviewer: I miss an explanation of how the build cache is different from incremental builds. I think this part should cover both and start with the incremental builds as the more basic optimization.

Author: Fixed in latest commit.

> Task :app:startScripts UP-TO-DATE
> Task :app:distTar UP-TO-DATE
> Task :app:distZip UP-TO-DATE
> Task :app:assemble UP-TO-DATE
> Task :app:compileTestJava UP-TO-DATE
> Task :app:processTestResources NO-SOURCE
> Task :app:testClasses UP-TO-DATE
> Task :app:test UP-TO-DATE
> Task :app:check UP-TO-DATE
> Task :app:build UP-TO-DATE
----

As we can see, Gradle used the _incremental build_ optimization.
Nothing has changed so Gradle has decided not to re-run some tasks.
Instead letting us know that inputs and outputs have not changed using the _outcome label_ `UP-TO-DATE`.

Now let's delete all the files by running the clean and build back to back:
[source]
----
$ ./gradlew :app:clean :app:build

> Task :app:clean
> Task :app:compileJava FROM-CACHE
> Task :app:processResources
> Task :app:classes
> Task :app:jar
> Task :app:startScripts
> Task :app:distTar
> Task :app:distZip
> Task :app:assemble
> Task :app:compileTestJava FROM-CACHE
> Task :app:processTestResources NO-SOURCE
> Task :app:testClasses UP-TO-DATE
> Task :app:test FROM-CACHE
> Task :app:check UP-TO-DATE
> Task :app:build

BUILD SUCCESSFUL in 551ms
9 actionable tasks: 6 executed, 3 from cache
----

> Nothing has changed but files were deleted during the clean task so Gradle has to re-run tasks to re-build the App.

Reviewer 2: ```suggestion
Nothing has changed in the project's source, but because build output files were deleted during the clean task Gradle has to re-run tasks to to re-created those files when re-building the App.
```

> Gradle is efficient, especially with the local cache turn on.

Reviewer 2: ```suggestion
Gradle is efficient, especially with the local cache turned on.
```

Gradle will look at the cache directory on your machine to check for output files that may already exist. If they do, instead of running that task, it will copy its (output) results into your project `build` directory.

The _outcome label_ `FROM-CACHE` lets the user know that Gradle has fetched the task results from the local cache.

|===
|Outcome Label |Description

|UP-TO-DATE
|Task that has been already executed and hasn't changed (incremental build feature)

|SKIPPED

> |Task has been skipped for some reason

Reviewer 2: "for some reason" doesn't sound good.  How about "Task was explicitly prevented from running."

|FROM-CACHE
|Task output has been copied to local directory from previous builds in the build cache (caching feature)

|NO-SOURCE

> |Task is unexecuted because the input data required for its execution was not found

Reviewer 2: ```suggestion
|Task was not executed because its required inputs were not available
```

-rw-r--r--  1 lkassovic  staff  3373 May 25 18:48 5d1e0e63f2ba1b5e6797d4bf186d8fa6
-rw-r--r--  1 lkassovic  staff  4217 May 25 18:48 e0ea15b80c7f67fdcd8b2902d1802da6
-rw-r--r--  1 lkassovic  staff     0 May 25 18:48 gc.properties
-rw-r--r--  1 lkassovic  staff    17 May 25 18:55 build-cache-1.lock
----

Here we see the cache key `7e6149c74419964bc38b387f6639a655`.
[source]
----
$ ls -ltr ~/.gradle/caches/build-cache-1/7e6149c74419964bc38b387f6639a655

total 40
-rw-r--r--  1 lkassovic  staff  3822 May 25 18:48 7e6149c74419964bc38b387f6639a655
-rw-r--r--  1 lkassovic  staff  3373 May 25 18:48 5d1e0e63f2ba1b5e6797d4bf186d8fa6
-rw-r--r--  1 lkassovic  staff  4217 May 25 18:48 e0ea15b80c7f67fdcd8b2902d1802da6
-rw-r--r--  1 lkassovic  staff     0 May 25 18:48 gc.properties
-rw-r--r--  1 lkassovic  staff    17 May 25 18:55 build-cache-1.lock
----
`7e6149c74419964bc38b387f6639a655` is a TAR file.

When the file is decompressed, we can see it contains our `App` class:

image::tutorial/mac-tar-cache-exp.png[]

Gradle will periodically clean-up the local cache directory by removing entries that have not been used recently to conserve disk space.

== Step 5. Remote Caching

> The remote cache (called a Build Cache Node) is a shared build cache available via HTTP.

Reviewer 2: I think the "Node" terminology is GE-specific and HTTP is too specific here.  I'd just say something like:

"In addition to the local cache present on a single developer's own machine, Gradle can make use of remote caches stored elsewhere, which can be used by multiple developers."
